packages/vitest/src/node/reporters/json.ts

```diff
@@ -60,20 +84,74 @@ export class JsonReporter implements Reporter {
 
     const success = numFailedTestSuites === 0 && numFailedTests === 0
 
-    const testResults: Array<TestResult> = tests.map(t => ({
-      perfStats: {
-        runtime: t.result?.duration,
-        start: t.result?.startTime,
-        end: t.result?.duration && t.result?.startTime && t.result.duration + t.result.startTime,
-      },
-      displayName: t.name,
-      failureMessage: t.result?.error?.message,
-      skipped: t.mode === 'skip',
-      status: t.result?.state,
-      testFilePath: t.file?.filepath,
-    }))
-
-    const result: AggregatedResult = { numTotalTestSuites, numPassedTestSuites, numFailedTestSuites, numPendingTestSuites, numTotalTests, numPassedTests, numFailedTests, numPendingTests, numTodoTests, startTime: this.start, success, testResults }
+    const testResults: Array<FormattedTestResult> = []
+    const fileToTestCases = new Map<string, { file: File; tests: Test[] }>()
+
+    for (const test of tests) {
+      const file = test.file
+      if (file) {
+        if (!fileToTestCases.has(file.filepath))
+          fileToTestCases.set(file.filepath, { file, tests: [] })
+
+        fileToTestCases.get(file.filepath)!.tests.push(test)
+      }
+    }
+
+    for (const [filepath, { tests, file }] of fileToTestCases) {
+      let startTime = tests.reduce((prev, next) => Math.min(prev, next.result?.startTime ?? Infinity), Infinity)
+      if (startTime === Infinity)
+        startTime = this.start
+
+      const endTime = tests.reduce((prev, next) => Math.max(prev, (next.result?.startTime ?? 0) + (next.result?.duration ?? 0)), startTime)
+      const assertionResults = tests.map((t) => {
+        const ancestorTitles = [] as string[]
+        let iter: Suite | undefined = t.suite
+        while (iter) {
+          ancestorTitles.push(iter.name)
+          iter = iter.suite
+        }
+        ancestorTitles.reverse()
+
+        return {
+          ancestorTitles,
+          fullName: ancestorTitles.length > 0 ? `${ancestorTitles.join(' ')} ${t.name}` : t.name,
+          status: t.result != null ? StatusMap[t.result.state] : 'skipped',
+          title: t.name,
+          duration: t.result?.duration,
+          failureMessages: t.result?.error?.message == null ? [] : [t.result.error.message],
+        } as FormattedAssertionResult
+      })
+
+      testResults.push({
+        assertionResults,
+        startTime,
+        endTime,
+        status: tests.every(t =>
+          t.result?.state === 'pass'
+           || t.result?.state === 'skip'
+           || t.result?.state === 'run'
```

Comment:
Wouldn't be an internal error to still have a `run` state here?

Reply:
I'm not very familiar with the internal logic. Should I add a failure message in this case?

Reply:
Maybe leave it as failed?

Reply:
Should we throw an error here or append a failure message to the result? Or do anything to let the user know?

Reply:
I think we could emit a warning and set it as failed in the output json.

Reply:
done